Introduce XPU scope profiler extending existing XPU profiler plugin

moksiuc opened the pull request:
**Summary:**

As XPU became a PyTorch built-in device, the profiler support is indispensable part of functionality completeness. In this PR, the XPU scope profiler is introduced by extending existing XPU profiler plugin. The XPU scope profiler is built on the foundation of intel PTI toolkit (https://github.com/intel/pti-gpu), and underlying SYCL runtime. It allows to gather XPU hardware metrics. The LIBKINETO_NOXPUPTI option is used to enable or disable the whole XPU profiler plugin during kineto build stage.

**Changes:**
- Added new ActivityType : XPU_SCOPE_PROFILER, enabling the new scope profiler
- Outputs XPU hardware metrics from the new scope profiler in Perfetto counters display mode ("C")
- Added gtest

moksiuc committed:
scope profiler squashed

moksiuc committed:
Type fix

gujinghui commented on libkineto/src/plugin/xpupti/FindSYCLToolkit.cmake:
```diff
@@ -105,6 +105,5 @@ if((NOT SYCL_INCLUDE_DIR) OR (NOT SYCL_LIBRARY_DIR) OR (NOT SYCL_LIBRARY))
 endif()
 
 message(DEBUG "The SYCL compiler is ${SYCL_COMPILER}")
-message(DEBUG "The SYCL Flags are ${SYCL_FLAGS}")
```
Why remove this message?

moksiuc replied:
SYCL_FLAGS is not used in any of cmake files

gujinghui commented:
@moksiuc  It's great that we are going to update our PTI integration code, and introduce new profiler path.
Could you help address below questions?
1. The alternative of ScopeProfiler is the RangeProfiler for CUDA? Looks like the RangeProfler is not enabled in PyTorch by default so far. Do you know why?
2. This PR is too huge to review. Can we split it to several PRs? For example, one PR for code refactor or cleanup per kineto or PTI changes, one or two PRs for ScopeProfiler, one PR for ChromeTraceLogger enhancement, and add test cases for each PRs.
3. BTW, CUDA provides CUDA_DRIVER activity to trace the driver actions. We should provide L0 actions as the counterpart, right? I remember, PTI should be able to do that. Do we have plan to cover it?
https://github.com/pytorch/kineto/blob/1e30d37905ddc1f9bd5e8de8d5e1148b3689f918/libkineto/src/ActivityType.cpp#L29

moksiuc commented:
> 1. The alternative of ScopeProfiler is the RangeProfiler for CUDA? Looks like the RangeProfler is not enabled in PyTorch by default so far. Do you know why?
It is enabled by providing experimental_config=_ExperimentalConfig(...). I don't know why it is this way but we are enabling our profiler the same way. One of the reasons may be that Range/Scope profiler requires parameters like HW metrics names that are passed through _ExperimentalConfig.

moksiuc commented:
> 3. BTW, CUDA provides CUDA_DRIVER activity to trace the driver actions. We should provide L0 actions as the counterpart, right? I remember, PTI should be able to do that. Do we have plan to cover it?
>    https://github.com/pytorch/kineto/blob/1e30d37905ddc1f9bd5e8de8d5e1148b3689f918/libkineto/src/ActivityType.cpp#L29

For sure not in this PR. I'll add this to our list of tasks.

moksiuc committed:
Fixes

- removed rangeEnabled
- fix test to align to this removal
- erase used kernelActivity from map

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Code review fixes

- place of config initialization
- removal of passing unused C compiler flag into test cmake file

moksiuc committed:
Fix clang build on PT level

moksiuc committed:
Fix clang build on PT level 2nd approach

gujinghui reviewed (changes requested):
This PR is split to #1177, #1180, and more.

moksiuc committed:
lintrunner

moksiuc committed:
Improve json for presentation in perfetto

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Fix gtest after json update

moksiuc committed:
Put correct oneapi version in error msg

moksiuc committed:
Fix correct PTI and OneApi version

moksiuc commented:
@gujinghui this is the core of the scope profiler. When 2 smaller parts are merged this one would have only core profiler left.

sraikund16 committed:
lint

sraikund16 committed:
lint

moksiuc committed:
Fix merge error

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Move device gathering to enableScopeProfiler

Otherwise old profiler requires setting ZET_ENABLE_METRICS what is
incorrect requirement.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Resolve exception from desctuctor

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc commented:
@EikanWang, @gujinghui
Please review.
2 extracted parts are already merged.
Now this change contains only code directly related to scope profiler.

moksiuc commented:
Tests are run on pytorch part:
https://github.com/pytorch/pytorch/pull/165766

gujinghui reviewed (approved):
LGTM. I assume you already verified it on local real machine, right? @moksiuc

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Fix Win build: error C2039: 'back_inserter': is not a member of 'std'

moksiuc committed:
Update min PTI version error msg

moksiuc committed:
Remove unnecesary includes

gujinghui commented:
@sraikund16  could you help review this PR? Thanks.

divyanshk reviewed (commented):
Thanks for the PR. I'm ramping up on this space. I'm curious to understand how much does it differ from the existing XPU profiler - maybe it should exist as a separate class in itself ? How can I learn more about the scope profiler ?

Also, thank you for adding tests.

moksiuc commented:
> Thanks for the PR. I'm ramping up on this space. I'm curious to understand how much does it differ from the existing XPU profiler - maybe it should exist as a separate class in itself ? How can I learn more about the scope profiler ?
> 
> Also, thank you for adding tests.

This PR adds new activity ActivityType::XPU_SCOPE_PROFILER, accepted by XPU profiler. With this new mode enabled XPU profiler additionally gathers requested HW metrics.
I don't think the separate class is necessary as it just adds additional mode to a few existing so far.
The new mode is a XPU alternative for what is implemented in class CuptiRangeProfiler.

gujinghui commented:
> > Thanks for the PR. I'm ramping up on this space. I'm curious to understand how much does it differ from the existing XPU profiler - maybe it should exist as a separate class in itself ? How can I learn more about the scope profiler ?
> > Also, thank you for adding tests.
> 
> This PR adds new activity ActivityType::XPU_SCOPE_PROFILER, accepted by XPU profiler. With this new mode enabled XPU profiler additionally gathers requested HW metrics. I don't think the separate class is necessary as it just adds additional mode to a few existing so far. The new mode is a XPU alternative for what is implemented in class CuptiRangeProfiler.

Hi @divyanshk ,

For more details of CuptiRangeProfiler, please see this page. https://deepwiki.com/pytorch/kineto/3.1-nvidia-gpu-support-(cupti)#4-cupti-range-profiler-integration

@moksiuc is going to provide similar functionalities on XPU in this PR.

moksiuc committed:
Avoid adding getMetadata method to the interface

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Split classes to avoid excessive #ifdef macro usage

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

divyanshk reviewed (commented):
@moksiuc  @gujinghui Can you help me understand a few things, thanks
1.) Do we have a different intel PTI API that is being leveraged for the range profiler ? How stable and well-supported is that API ? I ask because CUPTIRangeProfiler uses a different CUPTI API (`CUPTI Profiling API`) from the canonical profiler. Also that API is marked for deprecation so we are thinking of deprecating and deleting the CUPTIRangeProfiler.
2.) Can you point me to the official documentation for scope profiler API for Intel?
3.) Deferring code restructuring concerns to @scotts, I am starting to think `ActivityType` might be over-used here; because activitytype should be signifying profiling activities/events but `CUDA_PROFILER_RANGE` and `XPU_SCOPE_PROFILER` are independent profilers.  Thoughts?

moksiuc committed:
resolve merge commit issues

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Fix win build

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Fix win build

moksiuc commented:
@scotts, could you review restructured code, please ?

moksiuc commented:
Whole cleanup extracted to separate PR in case it is preferred:
https://github.com/pytorch/kineto/pull/1256

moksiuc committed:
Revert conflicting files

moksiuc committed:
Rename files 0->V2, V1->0

moksiuc committed:
Rename 0->V2, V1->0 inside files

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Resolve merge conflict

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Apply review comments

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
unformat

moksiuc committed:
unformat

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityProfilerSession.cpp:
```diff
@@ -44,7 +44,7 @@ std::unordered_set<std::string_view>
 // =========== Session Constructor ============= //
 XpuptiActivityProfilerSession::XpuptiActivityProfilerSession(
-    XpuptiActivityApi& xpti,
+    SELECT_VERSION(XpuptiActivityApi) & xpti,
```
Can we also pass the ref of XpuptiActivityApi obj, but do dynamic_cast for XpuptiActivityApi to XpuptiActivityApiV2 if needed in the implementation?

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityProfilerSession.h:
```diff
@@ -23,17 +23,16 @@
-                                const libkineto::Config& config,
-                                const std::set<ActivityType>& activity_types);
+  XpuptiActivityProfilerSession(
+      SELECT_VERSION(XpuptiActivityApi) & xpti,
```
Can we also pass the ref of XpuptiActivityApi obj, but do dynamic_cast for XpuptiActivityApi to XpuptiActivityApiV2 if needed in the implementation?

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityProfilerSession.h:
```diff
@@ -111,13 +110,34 @@ class XpuptiActivityProfilerSession : public libkineto::IActivityProfilerSession
   libkineto::getLinkedActivityCallback cpuActivity_;
 
-  XpuptiActivityApi& xpti_;
+  SELECT_VERSION(XpuptiActivityApi) & xpti_;
```
ditto.

moksiuc committed:
Apply review comments - SELECT_VERSION macro

moksiuc committed:
Revert "unformat"

This reverts commit 52bb30a7ce97de8d9bb2d69ecfac1760de8ff0b5.

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityProfilerSession.h:
```diff
@@ -23,17 +23,16 @@
-                                const libkineto::Config& config,
-                                const std::set<ActivityType>& activity_types);
+  XpuptiActivityProfilerSession(
+      XpuptiActivityApiT& xpti,
```
Here we are passing the ref of obj. The original API works for both parent and derived objects?

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityProfilerSession.h:
```diff
@@ -111,13 +110,34 @@ class XpuptiActivityProfilerSession : public libkineto::IActivityProfilerSession
   libkineto::getLinkedActivityCallback cpuActivity_;
 
-  XpuptiActivityApi& xpti_;
+  XpuptiActivityApiT& xpti_;
```
As we keep the ref of xpti here, I think we can always claim it's a XpuptiActivityApi obj. Then, no change is needed here.

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiProfilerMacros.h:
```diff
@@ -32,4 +32,12 @@ using namespace libkineto;
+#else
+class XpuptiActivityApi;
+using XpuptiActivityApiT = XpuptiActivityApi;
+#endif
```
not clear to me... why we have to use the macro to redefine the API at build stage...

moksiuc replied:
It is not a macro. But anyway it is no longer necessary.

moksiuc committed:
No ApiT type alias

moksiuc committed:
Revert "unformat"

This reverts commit 7b7bea9ec80586cc5c98ccd1c81caa763af70daf.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
unformat

moksiuc committed:
unfix compilation errors

moksiuc commented on libkineto/src/plugin/xpupti/XpuptiActivityProfiler.cpp:
```diff
@@ -66,8 +66,14 @@ std::unique_ptr<libkineto::IActivityProfilerSession> XPUActivityProfiler::
+#endif
+
+  return std::make_unique<XpuptiActivityProfilerSessionT>(
+      XpuptiActivityApiT::singleton(), name(), config, activity_types);
```
Done

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityApi.cpp:
```diff
@@ -14,11 +14,6 @@ namespace KINETO_NAMESPACE {
-XpuptiActivityApi& XpuptiActivityApi::singleton() {
-  static XpuptiActivityApi instance;
-  return instance;
-}
```
why we move this func to V2 cpp file?

moksiuc committed:
Apply review comments

moksiuc committed:
Revert "unfix compilation errors"

This reverts commit d7b330fbbd6131d5117cc7320d4be51b4273700c.

moksiuc committed:
Revert "unformat"

This reverts commit 71d22cdfac7031cc215e56b18677ffbce03072a0.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit a139c2968403fd5849b6a492311ce064274df18c.

moksiuc committed:
Reapply "unfix compilation errors"

This reverts commit 493b5657770c3f6ce25f5e38cb9a9dc7e75e6b18.

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityApiV2.cpp:
```diff
@@ -0,0 +1,208 @@
+
+#if PTI_VERSION_AT_LEAST(0, 15)
+
+XpuptiActivityApi& XpuptiActivityApi::singleton() {
```
Sounds like the code is wrong. The original design of singleton is not good. Why we need to use singleton() to call the member function, or to access to member data here?

moksiuc committed:
Clean up Api singleton usage

moksiuc committed:
Revert "unfix compilation errors"

This reverts commit 86d3759a0e51f13683467a83ea6e335d5aab397b.

moksiuc committed:
Revert "unformat"

This reverts commit e2b0bf712be4123dacdfa831e266a1355a5a450b.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit da9b441b5bf3fcce52dc9d944a86069c53f90295.

moksiuc committed:
Reapply "unfix compilation errors"

This reverts commit 446f2533f61cdcfe9ff544955b1d23b74af35ea9.

moksiuc committed:
Separate Api and ApiV2 classes and their singletons

moksiuc committed:
Revert "unfix compilation errors"

This reverts commit fcd32232c2c0b800541ceaa5af23a3d378e0f8e4.

moksiuc committed:
Revert "unformat"

This reverts commit f420f821375bcf8605a771f01cfa626fa8b95a23.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 6e0717a8c82b78e92e8265b5345ef5e043129c36.

moksiuc committed:
Reapply "unfix compilation errors"

This reverts commit 3ae459cdac1f769568cfb233595dc3cf27603437.

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityApi.h:
```diff
@@ -32,8 +32,8 @@ class XpuptiActivityApi {
   static XpuptiActivityApi& singleton();
 
-  static void pushCorrelationID(int id, CorrelationFlowType type);
-  static void popCorrelationID(CorrelationFlowType type);
```
Looks like cupti and other pti have same static APIs. If the change is not necessary, let's revert it.

gujinghui replied:
Since V1 and V2 will be handled separately, we can revert all changes to the V1 .h and .cpp files because they are not necessary for this PR. This will make the PR clearer and easier to review.

moksiuc commented on libkineto/src/plugin/xpupti/XpuptiActivityProfiler.cpp:
```diff
@@ -66,8 +66,17 @@ std::unique_ptr<libkineto::IActivityProfilerSession> XPUActivityProfiler::
+#if PTI_VERSION_AT_LEAST(0, 15)
+  return std::make_unique<XpuptiActivityProfilerSessionV2>(
+      XpuptiActivityApi::singleton(),
+      XpuptiActivityApiV2::singleton(),
```
Let's try

moksiuc committed:
Rename XpuptiActivityApiV2 to XpuptiScopeProfilerApi

moksiuc committed:
Revert "Reapply "unfix compilation errors""

This reverts commit c2f9aed7d1fac462077a61da1684960298114248.

moksiuc committed:
Revert "Reapply "unformat""

This reverts commit 1f195d4b0404c11360f9628cd6b518c64619a47e.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit dc6c1f04978e6c38df89ab358ba4db01ad34c113.

moksiuc committed:
Reapply "unfix compilation errors"

This reverts commit 4483c756ff4d897605ad10e142ce799af13023e3.

moksiuc committed:
Fix merge conflict

moksiuc committed:
Revert "unfix compilation errors"

This reverts commit 481ea86628cf031a5587f51ba5c3426de32a6f7f.

moksiuc committed:
Revert "unformat"

This reverts commit d38f49f88e5487cabc407449f6b1f7677ca90dbb.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 267b4c060df40a2c88e708934acdfef9081ac43c.

moksiuc committed:
Reapply "unfix compilation errors"

This reverts commit 1e02e7580112b2fcb14e58ff181f7aea111d34ab.

gujinghui reviewed (approved):
@moksiuc  please resolve the change conflicts. Thanks.

moksiuc committed:
Revert "unfix compilation errors"

This reverts commit d2d99dec0003bde319a9ce07e8ca73b9782cb803.

moksiuc committed:
Revert "unformat"

This reverts commit 75cdaa06e8b627f0872c5cba17d7c97079da274a.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit f87ef7ac29e8a7984279724b656fa865056740ac.

gujinghui commented on libkineto/src/plugin/xpupti/XpuptiActivityApi.cpp:
```diff
@@ -244,6 +244,15 @@ void XpuptiActivityApi::enableXpuptiActivities(
+      case ActivityType::XPU_SCOPE_PROFILER:
+#if PTI_VERSION_AT_LEAST(0, 15)
+#else
+        throw std::runtime_error(
```
When does this error occur?

moksiuc committed:
Make ApiV2 a member of SessionV2 instead singleton

moksiuc committed:
Revert "unformat"

This reverts commit 6ce66520930f73d3dae447bd3cf642ad55d0670a.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit f55b81c8a7f16105afd5d877ecb97ad3b1b2e1d8.

moksiuc committed:
format

moksiuc committed:
Revert "unformat"

This reverts commit 4943d7211220428eba6ce3e3813c10bce21216d3.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 379d0774f926f97908e55cfc876c449fc4c3bddf.
This reverts commit 2a594f0ca529151bda7dffa2a592d98117847014.

divyanshk commented on libkineto/test/xpupti/XpuptiTestUtilities.cpp:
```diff
@@ -226,9 +226,33 @@ RunProfilerTest(
+        if (nameStartsWithMetrics)
+          goto label_scope;
+        else
+          goto label_default;
```
Let's avoid goto!

divyanshk commented on libkineto/libkineto_defs.bzl:
```diff
@@ -49,9 +49,12 @@ def get_libkineto_xpupti_srcs(with_api = True):
     return [
         "src/plugin/xpupti/XpuptiActivityApi.cpp",
         "src/plugin/xpupti/XpuptiActivityHandlers.cpp",
+        "src/plugin/xpupti/XpuptiActivityHandlersV2.cpp",
```
@scotts you had concerns on `xxx.cpp` and `xxxV2.xpp`

divyanshk commented on libkineto/src/plugin/xpupti/XpuptiActivityApi.cpp:
```diff
@@ -245,6 +245,15 @@ void XpuptiActivityApi::enableXpuptiActivities(
         break;
 
+      case ActivityType::XPU_SCOPE_PROFILER:
+#if PTI_VERSION_AT_LEAST(0, 15)
```
Nit: this is a little confusing to read, maybe just a comment on what happens on the if condition and hows its handled.

divyanshk reviewed (commented):
Left some comments. Important one is in `ActivityType.h`

I need to look again at `output_json.cpp`.

gujinghui commented on libkineto/include/ActivityType.h:
```diff
@@ -43,14 +43,15 @@ enum class ActivityType {
-  PRIVATEUSE1_RUNTIME = 24, // host side privateUse1 runtime events
-  PRIVATEUSE1_DRIVER = 25, // host side privateUse1 driver events
+  PRIVATEUSE1_RUNTIME = 25, // host side privateUse1 runtime events
+  PRIVATEUSE1_DRIVER = 26, // host side privateUse1 driver events
```
@moksiuc  Let's put the new item at the end of the list if still needed.

moksiuc committed:
address review comments

moksiuc committed:
Revert "unformat"

This reverts commit 2fe719ab3900366ff230794101199d42178a1085.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 60751c077930a31bff9a15cc5e1434a2bf78a3a4.

moksiuc committed:
Rename V2 -> Scope according to review comments

moksiuc committed:
Revert "unformat"

This reverts commit bda80d483d7790adb50d5bd7d3f3ffc0e730788c.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 8d9b89975540cc91ffb21cbde70df40c7b1e63ec.

moksiuc committed:
Align to new kineto code in output_json.cpp

moksiuc committed:
Revert "Reapply "unformat""

This reverts commit d0a82c14fd6d681156df11c65932a914dcc2c883.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 52300426dd5bcbab07e89427788cbcebee493c52.

gujinghui commented:
@divyanshk  @scotts all comments have been addressed. Could you help review again? Thanks.

moksiuc committed:
Revert "unformat"

This reverts commit f329732bb6f22e62d466d445e0d708a5eadc5b3e.

moksiuc committed:
Merge branch 'refs/heads/main' into moksiuci_6674_scope_profiler

moksiuc committed:
Reapply "unformat"

This reverts commit 3b2cb5a3df55945f659307f6ccf235ce37bb3632.

divyanshk reviewed (approved):
Thanks for addressing the comments! Looks good to me!

moksiuc commented:
Failure in the test
Linux CUDA build and test / pr-test / linux-cuda-build-and-test (pull_request)
seems unrelated to my change:
=================================== FAILURES ===================================
_______________________ TestProfiler.test_dynamic_toggle _______________________
Traceback (most recent call last):
  File "/opt/conda/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
    yield
  File "/opt/conda/lib/python3.11/unittest/case.py", line 623, in run
    self._callTestMethod(testMethod)
  File "/opt/conda/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
    if method() is not None:
       ^^^^^^^^
  File "/pytorch/kineto/pytorch/torch/testing/_internal/common_utils.py", line 3448, in wrapper
    method(*args, **kwargs)
  File "/pytorch/kineto/pytorch/torch/testing/_internal/common_utils.py", line 1821, in wrapper
    fn(*args, **kwargs)
  File "/pytorch/kineto/pytorch/test/profiler/test_profiler.py", line 2312, in test_dynamic_toggle
    self.assertTrue(all(device not in e.name for e in p1.events()))
  File "/opt/conda/lib/python3.11/unittest/case.py", line 715, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true

gujinghui commented:
@scotts   Could you help review this PR again? We already addressed all comments. Thanks.

aostrowski-hbn commented:
@scotts @divyanshk Can we merge?
